# Exercise 02 -  OLAP Cubes - Roll Up and Drill Down

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [1]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

## Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [5]:
%%time
%%sql
SELECT  dimDate.day, dimMovie.rating, dimCustomer.country,sum(sales_amount) as revenue
FROM factsales
JOIN dimMovie on (dimMovie.movie_key = factsales.movie_key)
JOIN dimDate on (dimDate.date_key = factsales.date_key)
JOIN dimCustomer on(dimCustomer.customer_key = factsales.movie_key)
GROUP BY (dimDate.day, dimMovie.rating, dimCustomer.country)
ORDER BY revenue desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.75 ms, sys: 129 µs, total: 4.88 ms
Wall time: 19.9 ms


[(30, 'G', 'India', Decimal('128.76')),
 (30, 'R', 'India', Decimal('95.78')),
 (30, 'PG-13', 'India', Decimal('94.83')),
 (27, 'PG-13', 'India', Decimal('90.85')),
 (30, 'NC-17', 'China', Decimal('87.76')),
 (30, 'G', 'Russian Federation', Decimal('83.84')),
 (21, 'PG-13', 'India', Decimal('82.86')),
 (18, 'NC-17', 'China', Decimal('82.84')),
 (30, 'G', 'Brazil', Decimal('81.85')),
 (20, 'NC-17', 'China', Decimal('76.84')),
 (30, 'PG-13', 'China', Decimal('76.84')),
 (18, 'R', 'India', Decimal('75.81')),
 (30, 'G', 'China', Decimal('73.83')),
 (29, 'NC-17', 'India', Decimal('72.85')),
 (20, 'R', 'India', Decimal('72.83')),
 (29, 'G', 'Russian Federation', Decimal('71.86')),
 (19, 'NC-17', 'China', Decimal('71.83')),
 (27, 'PG-13', 'Japan', Decimal('69.83')),
 (30, 'R', 'Mexico', Decimal('69.83')),
 (19, 'G', 'Brazil', Decimal('67.88'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

## Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [7]:
%%time
%%sql
SELECT  dimDate.day, dimMovie.rating, dimCustomer.district,sum(sales_amount) as revenue
FROM factsales
JOIN dimMovie on (dimMovie.movie_key = factsales.movie_key)
JOIN dimDate on (dimDate.date_key = factsales.date_key)
JOIN dimCustomer on(dimCustomer.customer_key = factsales.movie_key)
GROUP BY (dimDate.day, dimMovie.rating, dimCustomer.district)
ORDER BY revenue desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.37 ms, sys: 170 µs, total: 4.54 ms
Wall time: 21.8 ms


[(30, 'G', 'So Paulo', Decimal('66.90')),
 (30, 'PG-13', 'England', Decimal('47.93')),
 (19, 'G', 'Smolensk', Decimal('42.94')),
 (19, 'PG-13', 'Antofagasta', Decimal('40.94')),
 (30, 'R', 'Karnataka', Decimal('39.94')),
 (21, 'PG-13', 'Tamil Nadu', Decimal('37.94')),
 (30, 'NC-17', 'Coahuila de Zaragoza', Decimal('37.94')),
 (20, 'PG-13', 'Central Mindanao', Decimal('35.96')),
 (30, 'R', 'Songkhla', Decimal('35.94')),
 (30, 'PG-13', 'Tamil Nadu', Decimal('35.93')),
 (30, 'R', 'Gois', Decimal('34.95')),
 (19, 'G', 'So Paulo', Decimal('34.94')),
 (18, 'G', 'Buenos Aires', Decimal('34.93')),
 (29, 'PG-13', 'Campeche', Decimal('33.96')),
 (20, 'G', 'Punjab', Decimal('33.95')),
 (30, 'PG', 'Kalimantan Barat', Decimal('32.96')),
 (20, 'PG-13', 'Ninawa', Decimal('31.96')),
 (30, 'PG', 'Rio Grande do Sul', Decimal('31.95')),
 (18, 'PG', 'Adana', Decimal('31.95')),
 (19, 'PG-13', 'Tatarstan', Decimal('30.97'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>